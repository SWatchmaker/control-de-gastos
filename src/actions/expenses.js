import database from "../firebase/firebase";

//ADD_EXPENSE

const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(addExpense({ id: ref.key, ...expense }));
      })
      .catch(() => {
        console.log("error");
      });
  };
};

//REMOVE_EXPENSE

const removeExpense = ({ id = 0 } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

const startRemoveExpense = ({ id }) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      })
      .catch(() => {
        console.log("error");
      });
  };
};
//EDIT_EXPENSE

const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

const startEditExpense = (id, updates) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};
//SET EXPENSE
const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref("expenses")
      .once("value")
      .then(snapshot => {
        let expenses = [];
        snapshot.forEach(child => {
          expenses.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setExpenses(expenses));
      })
      .catch(() => {
        console.log("error");
      });
  };
};

export {
  addExpense,
  startAddExpense,
  removeExpense,
  startRemoveExpense,
  editExpense,
  startEditExpense,
  setExpenses,
  startSetExpenses
};
