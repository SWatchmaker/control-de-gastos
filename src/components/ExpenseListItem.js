import React from "react";
import { NavLink } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <NavLink to={"/edit/" + id}>{description}</NavLink>
    <p>
      {numeral(amount / 100).format("$0,0.00")} -
      {moment(createdAt).format("Do MMMM, YYYY")}
    </p>
  </div>
);

export default ExpenseListItem;
