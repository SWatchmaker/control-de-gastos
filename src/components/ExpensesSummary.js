import React from "react";
import getExpensesTotal from "../selectors/expenses-total";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import numeral from "numeral";

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => (
  <div>
    <h1>
      NÂ° Gastos: {expensesCount} | Total Gastos:
      {numeral(expensesTotal / 100).format("$0,0.00")}
    </h1>
  </div>
);

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesCount: visibleExpenses.length,
    expensesTotal: getExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
