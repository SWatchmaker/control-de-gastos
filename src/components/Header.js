import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
  <header>
    <h1>Control de Gastos</h1>
    <NavLink to="/" activeClassName="is-active" exact={true}>
      Dashboard
    </NavLink>
    <NavLink to="/create" activeClassName="is-active">
      AÃ±adir Gasto
    </NavLink>
    <NavLink to="/help" activeClassName="is-active">
      Ayuda
    </NavLink>
  </header>
);

export default Header;
