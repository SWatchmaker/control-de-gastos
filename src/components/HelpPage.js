import React from "react";

const HelpPage = () => <div>I need help!</div>;

export default HelpPage;
