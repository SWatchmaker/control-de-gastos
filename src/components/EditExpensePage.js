import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startEditExpense(expense);
    this.props.history.push("/");
  };

  onClick = expense => {
    this.props.startRemoveExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Editar Gasto</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onClick}>
            Borrar Gasto
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: expense => {
    dispatch(startEditExpense(props.match.params.id, expense));
  },
  startRemoveExpense: () => {
    dispatch(startRemoveExpense({ id: props.match.params.id }));
  }
});
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
