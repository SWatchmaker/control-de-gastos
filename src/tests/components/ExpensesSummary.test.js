import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import expenses from "../fixtures/expenses";

test("Renderiza resumen de gasto con 1 gasto.", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={1} expensesTotal={200} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("Renderiza resumen de gasto con mÃºltiples gastos.", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={21} expensesTotal={456700} />
  );
  expect(wrapper).toMatchSnapshot();
});
