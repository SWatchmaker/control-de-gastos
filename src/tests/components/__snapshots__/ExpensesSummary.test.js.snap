// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderiza resumen de gasto con 1 gasto. 1`] = `
<div>
  <h1>
    N° Gastos: 
    1
     | Total Gastos:
    $2.00
  </h1>
</div>
`;

exports[`Renderiza resumen de gasto con múltiples gastos. 1`] = `
<div>
  <h1>
    N° Gastos: 
    21
     | Total Gastos:
    $4,567.00
  </h1>
</div>
`;
